client/docs/plugin_model.md

```diff
@@ -0,0 +1,169 @@
+<!---
+Licensed under Creative Commons Attribution 4.0 International License
+https://creativecommons.org/licenses/by/4.0/
+--->
+
+# Building Contract Plugins #
+
+The plugins module provides a means to build and use operations that
+is independent (as much as possible) of the actual invocation of the
+operation. For example, the `mock-contract` plugin wraps the
+`inc_value` method in the contract object with an interface so that it
+can be invoked as part of a Python script, a pdo-shell script, or a
+bash script. Similarly, the `mint_tokens` command may invoke a series
+of operations across multiple contract objects in order to mint new
+NFT tokens.
+
+There are three general styles of plugins: *contract operations*,
+*contract commands* and *scripts*. Loosely, a contract operation wraps
+the invocation of a single method on a contract object. Contract
+operations are parameterized to interact with a specific contract
+object with defaults for enclave service and ledger
+connections. Contract commands provide a logical operation that spans
+multiple invocations (possibly across multiple contract
+objects). Scripts simply provide a useful local function (such as
+updating the current configuration). The PDO client library includes a
+basic set of scripts that should be sufficient for most uses.
+
+Finally, the plugins module provides a simple way to pass values
+between commands and operations to enable more complex scripts. A
+command may generate output or it may save a result that other
+commands can leverage. The composition becomes useful when writing
+complex scripts with multiple commands and operations using Python or
+the pdo-shell.
+
+## Plugin Structure ##
+
+While there is no particular need for separating the definition of
+contract operations and commands, it is often the case that plugins
+are built in parts: a set of operations on a contract object, a
+set of commands that implement logical operations that span multiple
+invocations on multiple objects, and rarely a set of utility scripts.
```

Comment:
Does this imply that in order to implement a command, a contract author MUST first implement the needed set of operations?

Reply:
no.